### PR TITLE
Add CacheEntryCleanerJob

### PR DESCRIPTION
**Pull request by mrnugget**

### CODE REVIEW

File: enterprise/internal/batches/store/batch_spec_execution_cache_entry.go [hunk 1]
```diff
@@ -153,6 +153,42 @@ func (s *Store) MarkUsedBatchSpecExecutionCacheEntry(ctx context.Context, id int
 	return s.Exec(ctx, q)
 }
 
+// cleanBatchSpecExecutionEntriesQuery collects cache entries to delete by
+// collecting enough so that if we were to delete them we'd be under
+// maxCacheSize again.
+const cleanBatchSpecExecutionEntriesQuery = `
```
Comment (eseliger) [inline enterprise/internal/batches/store/batch_spec_execution_cache_entry.go:159]: Our convention

```suggestion
const cleanBatchSpecExecutionEntriesQueryFmtstr = `
```

File: enterprise/internal/batches/store/batch_spec_execution_cache_entry.go [hunk 2]
```diff
@@ -153,6 +153,42 @@ func (s *Store) MarkUsedBatchSpecExecutionCacheEntry(ctx context.Context, id int
...
+  SELECT sum(octet_length(value)) AS total FROM batch_spec_execution_cache_entries
+),
+candidates AS (
+  SELECT
+    id
+  FROM (
+      SELECT entries.id,
+             entries.created_at,
+             entries.last_used_at,
+			 SUM(octet_length(entries.value)) OVER (ORDER BY COALESCE(entries.last_used_at, entries.created_at) ASC, entries.id ASC) AS running_size
```
Comment (eseliger) [inline enterprise/internal/batches/store/batch_spec_execution_cache_entry.go:171]: Weird indentation
```suggestion
      SELECT
        entries.id,
        entries.created_at,
        entries.last_used_at,
        SUM(octet_length(entries.value)) OVER (ORDER BY COALESCE(entries.last_used_at, entries.created_at) ASC, entries.id ASC) AS running_size
```

File: enterprise/internal/batches/store/batch_spec_execution_cache_entry.go [hunk 3]
```diff
@@ -153,6 +153,42 @@ func (s *Store) MarkUsedBatchSpecExecutionCacheEntry(ctx context.Context, id int
...
+      FROM batch_spec_execution_cache_entries entries
+      LEFT JOIN batch_spec_workspaces ON batch_spec_workspaces.batch_spec_execution_cache_entry_id = entries.id
+      WHERE batch_spec_workspaces.id IS NULL
+  ) t
+  WHERE
+    ((SELECT total FROM total_size) - t.running_size) >= %s
+)
+DELETE FROM batch_spec_execution_cache_entries WHERE id IN (SELECT id FROM candidates)
+;
```
Comment (eseliger) [inline enterprise/internal/batches/store/batch_spec_execution_cache_entry.go:180]:
```suggestion
```

File: enterprise/internal/batches/store/batch_spec_execution_cache_entry.go [hunk 4]
```diff
@@ -153,6 +153,42 @@ func (s *Store) MarkUsedBatchSpecExecutionCacheEntry(ctx context.Context, id int
...
+`
+
+func (s *Store) CleanBatchSpecExecutionCacheEntries(ctx context.Context, maxCacheSize int64) (err error) {
+	ctx, endObservation := s.operations.markUsedBatchSpecExecutionCacheEntry.With(ctx, &err, observation.Args{LogFields: []log.Field{
```
Comment (eseliger) [inline enterprise/internal/batches/store/batch_spec_execution_cache_entry.go:184]:
```suggestion
	ctx, endObservation := s.operations.cleanBatchSpecExecutionCacheEntries.With(ctx, &err, observation.Args{LogFields: []log.Field{
```

**Review by eseliger — APPROVED** [review]

**Comment by mrnugget** [comment]
@sourcegraph/batchers Erik and I paired on this, so I'm going to go ahead and merge it. If there's comments: leave them and I'll address post-merge!

**Pull request merged**